src/binary-writer.cc

```diff
@@ -97,10 +97,97 @@ struct RelocSection {
   std::vector<Reloc> relocations;
 };
 
-struct Symbol {
-  Index symbol_index;
-  SymbolType type;
-  Index element_index;
+class Symbol {
+ public:
+  struct Function {
+    static const SymbolType type = SymbolType::Function;
+    Index index;
+  };
+  struct Data {
+    static const SymbolType type = SymbolType::Data;
+    Index index;
+    Offset offset;
+    Address size;
+  };
+  struct Global {
+    static const SymbolType type = SymbolType::Global;
+    Index index;
+  };
+  struct Section {
+    static const SymbolType type = SymbolType::Section;
+    Index section;
+  };
+  struct Event {
+    static const SymbolType type = SymbolType::Event;
+    Index index;
+  };
+
+ private:
+  SymbolType type_;
+  std::string name_;
+  uint8_t flags_;
+  union {
+    Function function_;
+    Data data_;
+    Global global_;
+    Section section_;
+    Event event_;
+  };
+
+ public:
+  Symbol(const std::string& name, uint8_t flags, const Function& f)
+      : type_(Function::type), name_(name), flags_(flags), function_(f) {}
+  Symbol(const std::string& name, uint8_t flags, const Data& d)
+      : type_(Data::type), name_(name), flags_(flags), data_(d) {}
+  Symbol(const std::string& name, uint8_t flags, const Global& g)
+      : type_(Global::type), name_(name), flags_(flags), global_(g) {}
+  Symbol(const std::string& name, uint8_t flags, const Section& s)
+      : type_(Section::type), name_(name), flags_(flags), section_(s) {}
+  Symbol(const std::string& name, uint8_t flags, const Event& e)
+      : type_(Event::type), name_(name), flags_(flags), event_(e) {}
+
+  SymbolType type() const { return type_; }
+  const std::string& name() const { return name_; }
+  uint8_t flags() const { return flags_; }
+
+  SymbolVisibility visibility() const {
+    return static_cast<SymbolVisibility>(flags() & WABT_SYMBOL_MASK_VISIBILITY);
+  }
+  SymbolBinding binding() const {
+    return static_cast<SymbolBinding>(flags() & WABT_SYMBOL_MASK_BINDING);
+  }
+  bool undefined() const { return flags() & WABT_SYMBOL_FLAG_UNDEFINED; }
+  bool defined() const { return !undefined(); }
+  bool exported() const { return flags() & WABT_SYMBOL_FLAG_EXPORTED; }
+  bool explicit_name() const { return flags() & WABT_SYMBOL_FLAG_EXPLICIT_NAME; }
+  bool no_strip() const { return flags() & WABT_SYMBOL_FLAG_NO_STRIP; }
+
+  bool IsFunction() const { return type() == Function::type; }
```

Comment:
Why not just use `SymbolType::Function` here?  What purpose does the `Function::type` serve?

Reply:
The purpose of the `::type` definitions is to allow `InternSymbol` to know what kind of symbol to make.  I admit that I wasn't sure which name to use here in `IsFunction` and friends; if you prefer that I change it, I am happy to do so.

Reply:
I guess it will make more sense if/when we add data symbol but for now it looks like you could avoid a lot of complexity and avoid InternSymbol being a template at all by just passing the SymbolType to InternSymbol.

It seems like having a separate class nested inside of `Symbol`, for each different type of symbol payload adds a fair amount of complexity, but perhaps that is justified and will make more sense once the followups land?

lgtm aside from this minor concern.

Reply:
I see what you are saying, and if every symbol just had an `Index` payload it would be entirely clear-cut.  I guess since I saw the SymbolType::Data symbols having additional fields, I found it confusing for Symbol to have unused fields for e.g. SymbolType::Function symbols, which would propagate around as arguments to e.g. InternSymbol and related functions.  But that is a theoretical concern given that wabt doesn't know anything about data symbols, and my patches don't change that.

The followup #1537 doesn't do anything amazing here; this is just a question of data representation.  If you like I can change to a less structured `class Symbol { string name; uint8_t flags; SymbolType type; Index index; Offset offset; Address size; }` and assert that offset/size accessors only happen for data symbols (of which we will have none), or even exclude data symbols from the representation.  WDYT?

Reply:
I think I'm good with this approach.  What do you think binji?

Reply:
I'm happy either way, tbh. I like having more explicit types (i.e. more structured) since it seems like it will be easier to expand  if/when we add new symbol values. OTOH, YAGNI and all :)